### ISSUE TITLE
bug:  cannot received subscription message when MQTT broker restart.

### ISSUE THREAD

**Issue by tanxxjun321** [eclipse/paho.mqtt.golang#294]
- Mosquitto:              v1.5.5
- paho.mqtt.golang: v1.1.1
- os:                           ubuntu 16.04 LTS
- go:                           v1.9.4

I found my app cannot received subscription message when MQTT broker restart.
 
- #22   ` [feature] Re-subscribe on reconnect `
- #238 `Broker Restarts, client reconnects and subscribed methods do not work `

guess:

paho.mqtt.golang need send subscribed topics to MQTT broker when broker restart.

I copy some useful information:
- MQTT broker start:
``` shell
[megvii@tensorflow] ~/workspace/mosquitto/mosquitto-1.5.5
❯ mosquitto -v -c mosquitto.conf
1553773811: mosquitto version 1.5.5 starting
1553773811: Config loaded from mosquitto.conf.
1553773811: Opening ipv4 listen socket on port 1883.
1553773811: Opening ipv6 listen socket on port 1883.
1553773854: New connection from 10.201.102.129 on port 1883.
1553773854: New client connected from 10.201.102.129 as tanjun (c0, k30).
1553773854: Will message specified (13 bytes) (r0, q2).
1553773854:     camera/c3r-125/S_12345678/stop
1553773854: Sending CONNACK to tanjun (0, 0)
1553773854: Received SUBSCRIBE from tanjun
1553773854:     camera/c3r-125/+/check (QoS 2)
1553773854: tanjun 2 camera/c3r-125/+/check
1553773854: Sending SUBACK to tanjun
```
- MQTT broker restart:
``` shell
[megvii@tensorflow] ~/workspace/mosquitto/mosquitto-1.5.5
❯ mosquitto -v -c mosquitto.conf
1553773867: mosquitto version 1.5.5 starting
1553773867: Config loaded from mosquitto.conf.
1553773867: Opening ipv4 listen socket on port 1883.
1553773867: Opening ipv6 listen socket on port 1883.
1553773869: New connection from 10.201.102.129 on port 1883.
1553773869: New client connected from 10.201.102.129 as tanjun (c0, k30).
1553773869: Will message specified (13 bytes) (r0, q2).
1553773869:     camera/c3r-125/S_12345678/stop
1553773869: Sending CONNACK to tanjun (0, 0)
```

**Comment by tanxxjun321** [comment]
Add recover subscription when MQTT broker restart.

Resolves #294



details:
- client.go  line: 384 
``` Golang
380:	DEBUG.Println(CLI, "client is reconnected")
381: 	if c.options.OnConnect != nil {
382:		go c.options.OnConnect(c)
383:	}
384:
385: +	c.msgRouter.recoverRoute(c.oboundP)
386:
387:	c.workers.Add(4)
388:	go errorWatch(c)
389:	go alllogic(c)
390:	go outgoing(c)
391: 	go incoming(c)
392: }
```

- router.go line: 116 ( Qos should be record in router )
``` Golang
// recoverRoute starts recover all of subscription to mosquitto when reconnected mqtt-broker.
func (r *router) recoverRoute(c chan *PacketAndToken) {
	r.Lock()
	defer r.Unlock()
	for e := r.routes.Front(); e != nil; e = e.Next() {
		r := e.Value.(*route)

		token := newToken(packets.Subscribe).(*SubscribeToken)
		token.subs = append(token.subs, r.topic)

		sub := packets.NewControlPacket(packets.Subscribe).(*packets.SubscribePacket)
		sub.Topics = append(sub.Topics, r.topic)
		sub.Qoss = append(sub.Qoss, 0)

		c <- &PacketAndToken{p: sub, t: token}
	}
}
```

**Comment by alsm** [comment]
If the broker is restarted and has thrown away all its state it is expected that the subscriptions would no longer work, the broker does not know about them.
If this is a concern you have with your application you can use the OnConnect() call back to do your subscriptions, that is called whenever the client is reconnected.
